### ISSUE TITLE
Auto updating between patch/minor releases

### ISSUE THREAD

**Issue by RPiAwesomeness** [CactusDev/CactusBot#68]
Having the ability to have CactusBot receive notification of new releases (via GitHub webhooks or Sepal) and then having it download the new code seamlessly update itself would be very useful.

It would mainly be for patch releases, bug fixes and minor changes (example: x.x.0 -> x.x.1), but possibly for minor releases (x.2 -> x.3).

**Comment by 2Cubed** [comment]
This would be really cool.

We would have to take into consideration things that can't just be `git pull`ed, too - for example, changing a config file from a `.json` to a `.py`, like in v0.3 to v0.4.
Maybe we could release a script which updates from one version to the next? (So, to update from v0.3.5 to v0.4, it would run v0.3.5 to v0.3.6, and then v0.3.6 to v0.4?) This would be really helpful even for standard manual updating procedures. We could potentially add this file as a "binary" to each release?

**Comment by RPiAwesomeness** [comment]
@2Cubed Possibly, but it would preferable if this could be handled within the bot in a seamless manner, without requiring an external script.

**Comment by 2Cubed** [comment]
@RPiAwesomeness True. Not sure of any other way to accomplish this, though. :thinking:

**Comment by MysticalMage** [comment]
You could check the version at bot startup and tell the user that they can update with `python update.py` and have that run a separate updater script. That script would have to download the files (Possibly with [requests](http://docs.python-requests.org/en/master/)) and replace all files except itself, the configs, and the caches. Some care would have to go into updating old cache files possibly other things, but it could be done.

~Mage
